target/linux/ath79/files/drivers/gpio/rb4xx-gpio.c

```diff
@@ -0,0 +1,168 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * GPIO driver for the MikroTik RouterBoard 4xx series
+ *
+ * Copyright (C) 2008-2011 Gabor Juhos <juhosg@openwrt.org>
+ * Copyright (C) 2008 Imre Kaloz <kaloz@openwrt.org>
+ * Copyright (C) 2015 Bert Vermeulen <bert@biot.com>
+ * Copyright (C) 2020 Christopher Hill <ch6574@gmail.com>
+ *
+ * This file was based on the driver for Linux 2.6.22 published by
+ * MikroTik for their RouterBoard 4xx series devices.
+ */
+#include <linux/platform_device.h>
+#include <linux/gpio/driver.h>
+#include <linux/module.h>
+#include <linux/of_device.h>
+
+#include <mfd/rb4xx-cpld.h>
+
+struct rb4xx_gpio {
+	struct rb4xx_cpld *cpld;
+	struct device *dev;
+
+	struct gpio_chip chip;
+	struct mutex lock;
+	u16 values;		/* bitfield of GPIO 0-8 current values */
+};
+
+static int rb4xx_gpio_cpld_set(struct rb4xx_gpio *gpio, unsigned int offset,
+			       int value)
+{
+	struct rb4xx_cpld *cpld = gpio->cpld;
+	u16 values;
+	int ret;
+
+	mutex_lock(&gpio->lock);
+	values = gpio->values;
+
+	if (value)
+		values |= BIT(offset);
+	else
+		values &= ~(BIT(offset));
+
```

Comment:
if you still want to make sure no useless SPI chatter happens, you could add here a test for `values == gpio->values` which, if true, would set `ret = 0` and jump (`goto`) a label located right before `mutex_unlock()`.